shopinvader_api_payment/routers/payment.py

```diff
@@ -0,0 +1,184 @@
+    odoo_env: Annotated[api.Environment, Depends(odoo_env)],
+) -> TransactionProcessingValues:
+    """Create a payment transaction.
+
+    Input is data obtained from /payment/providers, with the provider selected by the
+    user. This route is public, so it is possible to pay anonymously.
+    """
+    if not payment_utils.check_access_token(
+        data.access_token,
+        data.payable,
+    ):
+        _logger.info("Invalid access token")
+        raise HTTPException(403)
+    payable_obj = Payable.model_validate(json.loads(data.payable))
+    # similar to Odoo's /payment/transaction route
+    if data.flow == "redirect":
+        providers_sudo = (
+            odoo_env["payment.provider"]
+            .sudo()
+            ._get_compatible_providers(
+                payable_obj.company_id,
+                payable_obj.partner_id,
+                payable_obj.amount,
+                currency_id=payable_obj.currency_id,
+            )
+        )
+        if not data.provider_id or data.provider_id not in providers_sudo.ids:
+            _logger.info(
+                "Invalid provider %s for partner %s",
+                data.provider_id,
+                payable_obj.partner_id,
+            )
+            raise HTTPException(403)
+        provider_sudo = odoo_env["payment.provider"].sudo().browse(data.provider_id)
+
+        # Create the transaction
+        transaction_values = odoo_env[
+            "shopinvader_api_payment.payment_router.helper"
+        ]._get_tx_create_values(data, provider_sudo, odoo_env)
+        tx_sudo = (
+            odoo_env["payment.transaction"]
+            .sudo()
+            .with_context(
+                shopinvader_api_payment=True,
+                shopinvader_api_payment_frontend_redirect_url=data.frontend_redirect_url,
+                shopinvader_api_payment_base_url=urljoin(
+                    str(request.url), "providers/"
+                ),
+            )
+            .create(transaction_values)
+        )
+        tx_sudo._log_sent_message()
+        return TransactionProcessingValues(
+            flow="redirect", **tx_sudo._get_processing_values()
+        )
+    else:
+        raise NotImplementedError("Only redirect flow is supported")
+
+
+class ShopinvaderApiPaymentRouterHelper(models.AbstractModel):
+    _name = "shopinvader_api_payment.payment_router.helper"
+    _description = "ShopInvader API Payment Router Helper"
+
+    def _get_additional_transaction_create_values(
+        self, data: TransactionCreate
+    ) -> dict:
+        # Intended to be extended for invoices, carts...
+        additional_transaction_create_values = {}
+        return additional_transaction_create_values
+
+    def _get_tx_create_values(
+        self,
+        data: TransactionCreate,
+        provider_sudo: PaymentProvider,
+        odoo_env: Annotated[api.Environment, Depends(odoo_env)],
+    ) -> dict:
+        payable_obj = Payable.model_validate(json.loads(data.payable))
```

Comment:
maybe this function should accept payable_obj as argument, to avoid reparsing the payable multiple times.

Reply:
I've thought about it, but we still need `data` for some other parameters, such as provider.